### DISCUSSION TITLE
policy based on unstable component versions

### DISCUSSION THREAD

**Discussion by mieliespoor** [DependencyTrack/dependency-track#2516] [Q&A] [answered]
I had a look at the policy page and don't see an easy way to pick up pre-release / unstable component versions.

Say a component has the version `1.0.0-rc1`, or `1.0.0-develop.15`, I want to be able to have a policy pick these components up. Doing this will make it easy for us to track projects deployed to production with unstable component dependencies.

**Comment by syalioune** [comment, marked as answer]
Hello @mieliespoor 

You can create a policy like the one below

![image](https://user-images.githubusercontent.com/6144741/221960698-9e95c604-c7ac-46a9-82e1-17e105336797.png)

It will capture component having matching version like `0.0.1-rc1`

![image](https://user-images.githubusercontent.com/6144741/221960944-1450a4e3-9041-4693-aea4-7b345382e579.png)

**Reply by mieliespoor** [reply]
oh that is very handy. I didn't know that you can use regex in here. I tried something but obviously it wasn't valid regex which is why it didn't work.

**Reply by mieliespoor** [reply]
OK, not quite working. See the snippets below. I tested it on the component page, but the same regex is applied on a policy and not showing anything after re-analysing a project. I will take a look at the logs to see if something is mentioned in the logs of the api.

My simple wip regex: https://regex101.com/r/BvDzhk/1

![image](https://user-images.githubusercontent.com/5008243/222087883-79d0be5b-22d2-4db8-97d7-ebe48de71e1c.png)

![image](https://user-images.githubusercontent.com/5008243/222088161-247ea9d8-828e-4420-8494-01ad27ee13ba.png)

![image](https://user-images.githubusercontent.com/5008243/222088491-97a1d6c6-9b97-42a0-87c2-42c313ef9bb0.png)

**Reply by mieliespoor** [reply]
Further tests... using this `.*-.*` in the version field on the component search bring back results (see below). Doing that in the policy doesn't result in any policy violations. Nothing in the logs - not sure if this will feature in debug logs...
![image](https://user-images.githubusercontent.com/5008243/222094303-ad6b5d78-6575-41bf-8cc1-3d0d88d3c855.png)

**Reply by syalioune** [reply]
> Nothing in the logs - not sure if this will feature in debug logs...

It won't. Looking at the code, the issue is with the comparison method which consider that null value does not match .* regex. You fall into that situation because your component group is empty. 
It will work if you modify your policy to either leave the group regex input empty or add another condition (same as before but with empty group regex)

**Reply by mieliespoor** [reply]
this might then safely fall into the category as being a bug. thank you for the response and feedback.

**Reply by syalioune** [reply]
the bug was spotted by someone else : see #2587

**Reply by walterdeboer** [reply]
That's an easy one :-) https://github.com/DependencyTrack/dependency-track/pull/2594

**Reply by mieliespoor** [reply]
resurrecting this as we are running 4.8.2 and this is not working. When I do the search on the component page, it works and I get back pre-release components. When I create a policy with exactly the same search, I haven't seen any policy violations yet.

Components page:
![image](https://github.com/DependencyTrack/dependency-track/assets/5008243/72a6a31a-0597-43a4-a600-29f957627363)

Policy: (note that I have tried both "Any" and "All" options for the operator field)
![image](https://github.com/DependencyTrack/dependency-track/assets/5008243/9d272d0e-82fc-4685-a15e-54036a21b60c)

**Reply by mieliespoor** [reply]
No I take that back. It seems that it starting to work now.

**Comment by walterdeboer** [comment]
In #2501 there is a method to test for stable or unstable versions. This could be used to create a simple policy... Maybe in the near future :-)

P.S. it looks for "-(dev|atlassian|preview|next|canary|snapshot|a|alpha|b|beta|rc|cr|m|mr|ea).*" in the label part of the version, i'll add "develop".

**Reply by mieliespoor** [reply]
thank you Walter. Your change will come in useful, will be eagerly waiting for it. Although I think the regex might be restrictive in certain cases, but it will work for most cases for us. Not sure if it will be possible to make this configurable. That would be great.

**Reply by walterdeboer** [reply]
I'm open for sugestions on aditional labels to include. :-) I could load the regex from the application.properties mabye, i'll have a look
